### PR TITLE
Update Daml Documentation to reflect command deduplication related changes [KVL-1094]

### PR DESCRIPTION
**Pull request by miklos-da**
Summary of changes:
* Introduced concept of change ID.
* Consistently use deduplication duration/period instead of deduplication time.

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: docs/source/app-dev/app-arch.rst
Comment (fabiotudone-da) [inline docs/source/app-dev/app-arch.rst:142]:
```suggestion
The ledger will then guarantee that further commands for the same application ID, submitting party and command ID will be ignored within the deduplication period.
```

Or `duplicated`

File: docs/source/app-dev/services.rst
Comment (fabiotudone-da) [inline docs/source/app-dev/services.rst:59]:
```suggestion
Each intended ledger change is identified by the **change ID**, consisting of the submitting :ref:`party <com.daml.ledger.api.v1.Commands.party>`, the :ref:`application ID <com.daml.ledger.api.v1.Commands.application_id>` and the :ref:`command ID <com.daml.ledger.api.v1.Commands.command_id>`.
```

File: docs/source/app-dev/services.rst
Comment (fabiotudone-da) [inline docs/source/app-dev/services.rst:61]:
```suggestion
The following application-specific IDs , all of which are included in completion events, can be set in commands:
```

File: docs/source/app-dev/services.rst
Comment (fabiotudone-da) [inline docs/source/app-dev/services.rst:63]:
```suggestion
- A :ref:`commandId <com.daml.ledger.api.v1.Commands.command_id>`, returned to the submitting application only; it can be used to correlate commands to completions.
```
I'd perhaps move it after the submission ID, which seems more important to me.

File: docs/source/app-dev/services.rst
Comment (fabiotudone-da) [inline docs/source/app-dev/services.rst:77]:
```suggestion
- A command submission is considered a duplicate submission if the ledger server receives the command within the deduplication period of a previous command with the same change ID.
```

File: docs/source/app-dev/app-arch.rst
Comment (andreaslochbihler-da) [inline docs/source/app-dev/app-arch.rst:142]: We can nowadays have several `actAs` parties.
```suggestion
The ledger will then guarantee that commands for the same application ID, submitting parties and command ID will be ignored within the deduplication period.
```

File: docs/source/app-dev/app-arch.rst
Comment (andreaslochbihler-da) [inline docs/source/app-dev/app-arch.rst:148]: Let's emphasize that applications should not change the deduplication time upon a resubmission. That way, the application cannot distinguish between the old behaviour (new submission falls within dedup time of the existing transaction) and the new behaviour (existing transaction falls in the dedup time of the new submission).
```suggestion
- If you are not sure whether a command was submitted successfully, just resubmit it with the same deduplication duration. If the new command was submitted within the deduplication period, the duplicate submission will safely be ignored. If the deduplication period has passed, you can assume the command was lost or rejected and a new submission is justified.
```

File: docs/source/app-dev/app-arch.rst
Comment (fabiotudone-da) [reply docs/source/app-dev/app-arch.rst:142]: @andreaslochbihler-da Can't we just say "for the same change ID"?

File: docs/source/app-dev/services.rst
Comment (andreaslochbihler-da) [inline docs/source/app-dev/services.rst:59]: We take `actAs union party` as the submitting parties. Do you want to change this here, too?

Comment (fabiotudone-da) [reply docs/source/app-dev/services.rst:59]: Is saying "the submitting parties" too vague?

File: docs/source/app-dev/services.rst
Comment (andreaslochbihler-da) [inline docs/source/app-dev/services.rst:77]: That's not really what's happening in Canton any more. Canton uses the new idea of backward-looking dedup, i.e., the dedup period of the later submission counts, not the one of the accepted submission. I'm not sure what ledger implementations currently still use the forward-looking semantics; kvutils has been changed so that it always uses the max dedup time, which makes the two indistinguishable.

**Review by andreaslochbihler-da — APPROVED** [review]
A step in the right direction. Introducing the change ID makes sense.

I don't have a strong opinion whether we already now want to document the new backward-looking semantics of the deduplication period. Do we still have supported ledgers that implement the forward-looking semantics?

File: docs/source/app-dev/app-arch.rst
Comment (andreaslochbihler-da) [reply docs/source/app-dev/app-arch.rst:142]: > Can't we just say "for the same change ID"?

Sure. Then we should just make sure that we've defined that term previously in the documentation. Can we put a link to the definition in `services.rst`?

File: docs/source/app-dev/services.rst
Comment (fabiotudone-da) [reply docs/source/app-dev/services.rst:77]: > I'm not sure what ledger implementations currently still use the forward-looking semantics

I'm not sure either, perhaps @bame-da knows more. In addition to that, I'm not sure if the end state is having the backward-looking semantics specified as part of the API, of if the precise behavior should be left unspecified at the API level.

File: docs/source/app-dev/services.rst
Comment (andreaslochbihler-da) [reply docs/source/app-dev/services.rst:59]: I find that the links to the corresponding protobuf fields help the reader. It is weird if two fields are linked and we just leave "submitting parties" unanchored.

File: docs/source/app-dev/services.rst
Comment (andreaslochbihler-da) [reply docs/source/app-dev/services.rst:77]: > the end state is having the backward-looking semantics specified as part of the API

Yes. That's the desired end state once we have all officially supported ledger implementations moved over.

File: docs/source/app-dev/app-arch.rst
Comment (fabiotudone-da) [reply docs/source/app-dev/app-arch.rst]: 8b7eec5

File: docs/source/app-dev/services.rst
Comment (fabiotudone-da) [reply docs/source/app-dev/services.rst]: 8b7eec5, let me know if it's clear or if you think that wording it directly as `actAs union party` with links is better.

File: docs/source/app-dev/app-arch.rst
Comment (fabiotudone-da) [reply docs/source/app-dev/app-arch.rst]: cbe4a1f

File: docs/source/app-dev/services.rst
Comment (fabiotudone-da) [reply docs/source/app-dev/services.rst]: 8b7eec5

File: docs/source/app-dev/services.rst
Comment (fabiotudone-da) [reply docs/source/app-dev/services.rst]: 8b7eec5

File: docs/source/app-dev/services.rst
Comment (fabiotudone-da) [reply docs/source/app-dev/services.rst]: How about a more general wording until we ensure that all ledgers comply with the new semantics:

```
- A command submission is considered a duplicate submission if the ledger API server is aware of another command within the deduplication period and with the same :ref`change ID <change-id>`.
 ```

Or, we could expand on the new semantics; I think you @bame-da can probably help deciding here.

File: docs/source/app-dev/services.rst
Comment (andreaslochbihler-da) [reply docs/source/app-dev/services.rst]: That's great. Thanks.

File: docs/source/app-dev/services.rst
Comment (andreaslochbihler-da) [reply docs/source/app-dev/services.rst]: I like your unspecific wording here. It's always fine to make stronger promises in later releases.

Comment (fabiotudone-da) [reply docs/source/app-dev/services.rst]: Alright, 4792a0a34be0a857ad7285ffda1d5d950a8c11a3. @bame-da (or someone else from product): I'll also wait for your feedback/stamp before merging.

**Review by hubert-da — COMMENTED** [review]
LGTM

**Comment by fabiotudone-da** [comment]
Merging after a nod from @bame-da.

**Pull request merged**
